### ISSUE TITLE
Question about RAM consumption increasing.

### ISSUE THREAD

**Issue by isaacvitor** [NVIDIA/NeMo#5281]
Hello everybody.

Forgive me for the stupid question but: 
Is expected the RAM consumption to increase or have high usage values?

I've been using a `stt_en_conformer_ctc_large.nemo` in GPU without any problem, to be more specific I'm running two models in two child processes. After every transcription, `torch.cuda.empty_cache()` solve my problems with the GPU cache and so on.

But I noticed when these processes start, the RAM had a significantly increased consumption (basically starting from model size + other things) and increase transcription after transcription.

That is the way I'm loading the models:
```python
 asr_model = nemo_asr.models.EncDecCTCModel.restore_from(_model_path)
 asr_model.eval()
 asr_model = asr_model.to(asr_model.device)
```

Is this normal behavior? Is there anything that I can do to solve it?

Thank you!

**Comment by titu1994** [comment]
Restore_from by default will have map_location=None which will check if you have a gpu or not and then load if in that directly. 

So if you want only one model on the GPU, you need to pass map_location="cpu" to restore from and then manually call asr_model.to("cuda:0") or other cuda I'd if you have multiple gpu. 

Also, eval alone is not good for inference you can use model.freeze() to ensure no gradient calculation is possible or use with torch.inference_mode(): 

Btw I don't think we've needed to call torch.cuda.empty_cache() cause it will slowdown inference by a lot. But you may keep it for safety sake. 

Can you show your transcribe call ? If you're using logprobs=True, then there's a high chance you're leaking memory from the returned tensor which is still in the gpu. Otherwise there's no reason subsequent transcribe calls should have higher memory.

**Comment by isaacvitor** [comment]
> 

Thank you for the tips! I forget to mention, I'm using Nemo version 1.11.0.
 
In a certain way, I'm using logprobs=True. I did a workaround to have `timestamps` following this https://github.com/NVIDIA/NeMo/discussions/2577#discussioncomment-1134049 discussion. This can be the source of my memory usage increase.

But I did a simple test, just loading the model as you suggested, something like this:
```python
asr_model = nemo_asr.models.EncDecCTCModel.restore_from(_model_path, map_location='cpu')
asr_model.to('cuda:0')
asr_model.freeze()
```
My ram and GPU before the load:
```bash
MiB Mem :  15748.9 total,   6497.3 free,   6964.6 used,   2287.1 buff/cache
```
```bash
GPU 0MiB / 15109MiB 
```
After
```bash
MiB Mem :  15748.9 total,   3729.1 free,   9720.4 used,   2299.5 buff/cache
```
```bash
GPU 1362MiB / 15109MiB 
```
Ram starts on 6964.6 and goes to 9720.4. What makes me confused is why the RAM is used in this process.

**Comment by titu1994** [comment]
Well if you're using CPU then your tensors are on CPU ram, so that is expected

**Comment by isaacvitor** [comment]
Sorry, probably I wasn't clear. 
I'm not using CPU but GPU, `map_location=cpu` I used just because of you suggest it.
Doesn't matter if I use `map_location=cuda[torch.cuda.device()]` or `model.to('cuda:0')` the CPU memory is used and it is the point.

**Comment by titu1994** [comment]
Model transcribe will force into gpu of it'd there. If you really want it to be on CPU only then you have to use CUDA_VISIBLE_DEVICES="" before calling python script to prevent use of gpu entirely. 

Model transcribe() has dataloaders that will default to GPU 0 for data processing if no device is provided for the model.

**Comment by titu1994** [comment]
Some CPU memory will always be used no matter whether you put your model on GPU or not. Pytorch execution has some cost, so does python environment and classes that run on it.

**Comment by titu1994** [comment]
You'll have to post your inference script to see if there are obvious reasons for memory leak. Not really possible to do so given model loading snippets.

**Comment by isaacvitor** [comment]
As I said

> Some CPU memory will always be used no matter whether you put your model on GPU or not. Pytorch execution has some cost, so does python environment and classes that run on it.

I understand it indeed, but my usage of CPU memory is greater than  GPU memory when I set `map_location=cuda`.

**Comment by titu1994** [comment]
You'll have to posts your inference script, I don't know how you're running multiple processes and why it's consuming more CPU memory.

I have another pr that launches multiple Nemo processes just fine and it has no CPU or GPU memory issue and can do concurrent training even. 

https://github.com/NVIDIA/NeMo/pull/5159

**Comment by titu1994** [comment]
Oh forgot, there's also the ASR webapp that runs on gunicorn with multiple workers just fine without oom or drastic CPU memory so something may be incorrect in your launch config or setup of the model 

https://github.com/NVIDIA/NeMo/tree/main/tools/asr_webapp

**Comment by isaacvitor** [comment]
As I said before my inference script could be the problem of the increasing memory. But to analyze the CPU memory consumption is not necessarily more than it:
```python
import nemo.collections.asr as nemo_asr
_model_path = 'model/stt_en_conformer_ctc_large.nemo'
 asr_model = nemo_asr.models.EncDecCTCModel.restore_from(_model_path, map_location='cuda')
 ```
 With this small snippet, I could notice an increase of 2755.8Mb in CPU Memory and this doesn't make sense.

**Comment by isaacvitor** [comment]
Basically my inference script is this:
```python
# all imports
class AudioDataLayer(IterableDataset):
    @property
    def output_types(self):
        return {
            'audio_signal': NeuralType(('B', 'T'), AudioSignal(freq=self._sample_rate)),
            'a_sig_length': NeuralType(tuple('B'), LengthsType()),
        }

    def __init__(self, sample_rate):
        super().__init__()
        self._sample_rate = sample_rate
        self.output = True

    def __iter__(self):
        return self

    def __next__(self):
        if not self.output:
            raise StopIteration
        self.output = False
        return self.signal, self.signal_shape

    def set_signal(self, signal):
        self.signal = signal
        self.signal_shape = torch.as_tensor(len(signal), dtype=torch.int64)
        self.output = True

    def __len__(self):
        return 1


def softmax(x):
    e = np.exp(x - np.max(x))
    return e / e.sum(axis=-1).reshape([x.shape[0], 1])


class FrameASR:
    """
    Based on:https://github.com/NVIDIA/NeMo/discussions/2577 and
    https://github.com/NVIDIA/NeMo/discussions/2577#discussioncomment-1134049 and
    https://github.com/parlance/ctcdecode/issues/114#issuecomment-961676918 (fixing bug)
    """

    def __init__(self, asr_model, sample_rate=SAMPLE_RATE, token_offset=TOKEN_OFFSET, time_step=TIME_STEP,
                 time_pad=TIME_PAD, frame_len=FRAME_LEN, lm_path=LM_PATH):
        '''
        Args:
        frame_len: frame's duration, seconds
        '''
        self.asr_model = asr_model

        self.sr = sample_rate
        self.frame_len = frame_len
        self.n_frame_len = int(frame_len * self.sr)

        self.data_layer = AudioDataLayer(sample_rate=self.asr_model.cfg.preprocessor.sample_rate)
        self.data_loader = DataLoader(self.data_layer, batch_size=1, collate_fn=self.data_layer.collate_fn)

        self.vocab = self.asr_model.decoder.vocabulary
        self.vocab.append("▁")

        self.time_step = time_step
        self.time_pad = time_pad

        self.beam_search_lm = CTCBeamDecoder(
            [chr(idx + token_offset) for idx in range(len(self.vocab))],
            model_path=lm_path,
            alpha=0.75,
            beta=1.0,
            beam_width=128,
            num_processes=max(os.cpu_count(), 1),
            blank_id=128,
            cutoff_prob=1.0,
            cutoff_top_n=40,
            log_probs_input=False)

    def _decode(self, frame, chunk_pad):
        logits = self.infer_signal(self.asr_model, frame).cpu().numpy()[0]
        decoded = self._beam_decoder(logits, chunk_pad)
        return decoded

    @torch.no_grad()
    def transcribe(self, frame=None, chunk_pad=0):
        if frame is None:
            frame = np.zeros(shape=self.n_frame_len, dtype=np.float32)
        return self._decode(frame, chunk_pad)

    def _beam_decoder(self, logits, chunk_pad):
        probs = softmax(logits)
        probs_seq = torch.FloatTensor(probs)

        beam_results, beam_scores, timesteps, out_lens = self.beam_search_lm.decode(probs_seq.unsqueeze(0))

        beam_res = beam_results[0][0][:out_lens[0][0]].cpu().numpy()
        times = timesteps[0][0][:out_lens[0][0]].cpu().numpy()
        lens = out_lens[0][0]

        transcript = self.asr_model.tokenizer.ids_to_text(beam_res)

        if len(times) > 0:
            if times[0] < self.time_pad:
                start = chunk_pad
                end = (times[0] + self.time_pad*2) * self.time_step + chunk_pad
            else:
                start = (times[0] - self.time_pad) * self.time_step + chunk_pad
                end = (times[0] + self.time_pad) * self.time_step + chunk_pad

            tocken_prev = self.vocab[int(beam_res[0])]
            word = tocken_prev

            result = []

            for n in range(1, lens):
                tocken = self.vocab[int(beam_res[n])]

                if tocken[0] == "#":
                    word = word + tocken[2:]

                elif tocken[0] == "-" or tocken_prev[0] == "-":
                    word = word + tocken

                else:
                    if start > end:
                        result_word = {"start": round(end, 3), "end": round(start, 3), "word": word}
                    else:
                        result_word = {"start": round(start, 3), "end": round(end, 3), "word": word}

                    result.append(result_word)

                    if times[n] < self.time_pad:
                        start = chunk_pad
                    else:
                        start = (times[n] - self.time_pad) * self.time_step + chunk_pad

                    word = tocken

                if times[n] < self.time_pad:
                    end = (times[n] + self.time_pad*2) * self.time_step + chunk_pad
                else:
                    end = (times[n] + self.time_pad) * self.time_step + chunk_pad

                tocken_prev = tocken

            if start > end:
                result_word = {"start": round(end, 3), "end": round(start, 3), "word": word}
            else:
                result_word = {"start": round(start, 3), "end": round(end, 3), "word": word}

            result.append(result_word)

        else:
            result = []

        return transcript, result

    def infer_signal(self, model, signal):
        self.data_layer.set_signal(signal)
        batch = next(iter(self.data_loader))
        audio_signal, audio_signal_len = batch
        audio_signal, audio_signal_len = audio_signal.to(
            self.asr_model.device), audio_signal_len.to(
            self.asr_model.device)
        log_probs, encoded_len, predictions = model.forward(
            input_signal=audio_signal, input_signal_length=audio_signal_len
        )
        return log_probs


def load_model(model_name=MODEL_NAME, model_path=MODEL_PATH, model_url=MODEL_URL):
    try:
        # Setting device
        device = 'cuda' if torch.cuda.is_available() else 'cpu'

        # Creating model path
        _model_path = os.path.join(model_path, model_name)


        # Setting model
        asr_model = nemo_asr.models.EncDecCTCModel.restore_from(_model_path, map_location=device)
        asr_model.freeze()

        return asr_model

    except Exception as e:
        logger.debug('Error while initializing Model')
        raise e


def create_recognizer(asr_model, frame_len=FRAME_LEN, lm_path=LM_PATH):
    return FrameASR(asr_model, frame_len=frame_len, lm_path=lm_path)


@torch.no_grad()
def transcribe(audio_path, frame_asr, sample_rate=SAMPLE_RATE, frame_len=FRAME_LEN, lm_path=LM_PATH):
    chunk_size = int(frame_len*sample_rate)
    file_bin_data = AudioSegment.from_file(audio_path, 'wav').set_frame_rate(sample_rate).get_array_of_samples()

    file_np = np.frombuffer(file_bin_data, dtype=np.int16)/32768.
    file_tensor = torch.FloatTensor(file_np)

    frame_asr = frame_asr

    sum_text = ""
    sum_res = []

    if len(file_tensor) > chunk_size:
        get_chunks = torch.tensor_split(file_tensor, math.ceil(len(file_tensor)/chunk_size))
        chunk_pad = 0
        for chunk in get_chunks:
            text, res = frame_asr.transcribe(frame=chunk, chunk_pad=chunk_pad/sample_rate)
            chunk_pad = chunk_pad + len(chunk)
            sum_text = sum_text + ' ' + text
            sum_res = sum_res + res

    elif len(file_tensor) > 0:
        text, res = frame_asr.transcribe(frame=file_tensor, chunk_pad=0)
        sum_text = sum_text + ' ' + text
        sum_res = sum_res + res

    sum_text = sum_text[1:].replace('  ', ' ').replace('#', '')
    torch.cuda.empty_cache()
    return sum_text, sum_res
```

**Comment by titu1994** [comment]
I haven't gone through your full code but it seems basically the buffered inference CTC code. 

I dunno why it's taking 2.7 GB of CPU memory.

**Comment by isaacvitor** [comment]
> I haven't gone through your full code but it seems basically the buffered inference CTC code.
> 
> I dunno why it's taking 2.7 GB of CPU memory.

Did you tried only the load model? The load model cause 2.7Gb  usage of CPU memory. This doesn't make sense.

**Comment by titu1994** [comment]
Yes I checked with htop, it temporarily increased to that size then went down after the model was moved to the GPU.

**Comment by isaacvitor** [comment]
> Yes I checked with htop, it temporarily increased to that size then went down after the model was moved to the GPU.
> 
> This is again expected - you aren't directly loading tensor into gpu memory, you first load a model into CPU then move it onto gpu.

This is strange, because in my case using only the REPL, after move CPU to GPU, CPU memory keeps high.

**Comment by titu1994** [comment]
I dunno then.

**Comment by isaacvitor** [comment]
@titu1994 could you share your load model?

**Comment by isaacvitor** [comment]
I will close this question because I'm convinced that the problem is pytroch.
The snippet below is enough to put the CPU memory usage high: 
```python
import torch 
a = torch.tensor([0], device="cuda")
```
For further reference: https://discuss.pytorch.org/t/cpu-ram-usage-with-cuda-is-large-2-gb/117668
